processing/src/main/java/org/apache/druid/frame/channel/ComposingReadableFrameChannel.java

```diff
@@ -0,0 +1,147 @@
+import java.util.function.Supplier;
+
+/**
+ * A composed readable channel to read frames. The channel can encapsulate multiple readable channels in it and
+ * automatically switches to next channels once the currently read channel is finished.
+ */
+public class ComposingReadableFrameChannel implements ReadableFrameChannel
+{
+  private final List<Supplier<ReadableFrameChannel>> channels;
+  private ReadableFrameChannel currentChannel;
+  private int currentIndex;
+
+  public ComposingReadableFrameChannel(
+      int partition,
+      List<Supplier<ReadableFrameChannel>> channels,
+      Map<Integer, HashSet<Integer>> partitionToChannelMap
+  )
+  {
+    Preconditions.checkNotNull(channels, "channels is null");
+    Preconditions.checkNotNull(partitionToChannelMap, "partitionToChannelMap is null");
+    if (partitionToChannelMap.get(partition) == null) {
+      // no writes for the partition, send an empty readable channel
+      this.channels = ImmutableList.of(() -> ReadableNilFrameChannel.INSTANCE);
+    } else {
+      HashSet<Integer> validChannels = partitionToChannelMap.get(partition);
+      Preconditions.checkState(validChannels.size() > 0, "No channels found for partition " + partition);
+      ImmutableList.Builder<Supplier<ReadableFrameChannel>> validChannelsBuilder = ImmutableList.builder();
+      ArrayList<Integer> sortedChannelIds = new ArrayList<>(validChannels);
+      Collections.sort(sortedChannelIds); // the data was written from lowest to highest channel
+      for (Integer channelId : sortedChannelIds) {
+        validChannelsBuilder.add(channels.get(channelId));
+      }
+      this.channels = validChannelsBuilder.build();
+    }
+    this.currentIndex = 0;
+    this.currentChannel = null;
+  }
+
+  @Override
+  public boolean isFinished()
+  {
+    initCurrentChannel();
+    if (!currentChannel.isFinished()) {
+      return false;
+    }
+    currentChannel.close();
+    currentChannel = null;
+    if (isLastIndex()) {
+      return true;
+    }
+    ++currentIndex;
+    return isFinished();
+  }
+
+  @Override
+  public boolean canRead()
+  {
+    initCurrentChannel();
+    if (currentChannel.canRead()) {
+      return true;
+    }
+    if (currentChannel.isFinished()) {
+      currentChannel.close();
+      currentChannel = null;
+      if (isLastIndex()) {
+        return false;
+      }
+      ++currentIndex;
+      return canRead();
+    }
+    return false;
+  }
+
+  @Override
+  public Frame read()
+  {
+    return currentChannel.read();
+  }
+
+  @Override
+  public ListenableFuture<?> readabilityFuture()
+  {
+    initCurrentChannel();
+    if (!currentChannel.isFinished()) {
+      return currentChannel.readabilityFuture();
+    }
+    currentChannel.close();
+    currentChannel = null;
+    if (isLastIndex()) {
+      return Futures.immediateFuture(true);
+    }
+    ++currentIndex;
+    return readabilityFuture();
+  }
+
+  @Override
+  public void close()
+  {
+    if (currentChannel != null) {
```

Comment:
Shouldnt we close all the "channels here"
`Releases any resources associated with this readable channel. After calling this, you should not call any other methods on the channel.`

Reply:
Same case as writable channels - the previous ones are closed before moving to next and the future ones are lazy.